### ISSUE TITLE
Support terminate!

### ISSUE THREAD

**Issue by ChrisRackauckas** [SciML/DiffEqGPU.jl#43]
We will need to think about this one. Actually terminating will cause the others not to finish. I think that terminate should just set an array to remember the time, and then we truncate back to the terminated times after solving.

**Comment by aml5600** [comment]
Could this support synchronizing any arbitrary callback?

**Comment by ChrisRackauckas** [comment]
it could, but it would need to make sure all want to terminate.

**Comment by aml5600** [comment]
sorry sorry, synchronize any "affect" of a callback, instead of just terminating.

**Comment by ChrisRackauckas** [comment]
what do you mean?

**Comment by aml5600** [comment]
In the case of a possibly hybrid system, being run on a GPU, it makes sense to synchronize the instances at the switching point so only one subsystem is executed at a time. Although I guess this would also require more than one generated kernel...obviously this can be circumvented with `du = (use_a)*sys_a(u) + (!use_a)*sys_b(u)`. 

Maybe this is a too targeted application though.

**Comment by ChrisRackauckas** [comment]
Running both is probably much faster on the GPU, and this is unrelated to this issue.

**Comment by fjebaker** [comment]
@ChrisRackauckas I have a proof of concept for how this could work with the new GPUTsit5 integrators: since they have one integrator per thread, it's very easy to track state and modify it when `terminate!` is called, in turn breaking out of the integration loop on that thread.

I tried taking a look at the DiffEqBase callback code to see if I could tie this in with the rest of the callback handling, but at the moment there's a few too many inconsistencies with GPUTsit5 vs. regular integrators (e.g. not having `opts`, and other missing fields) to just call the DiffEqBase methods (though someone with more insight could probably do it). As a work-around, I mimicked some of the callback functionality in a `gpu_apply_callbacks!` function. 

I can tidy things up and push my changes to a fork if you think there's value in this approach?

--

Also, for the ensemble array methods, I don't think there's a nice way of _actually_ terminating a single integration. E.g., say you have lots of trajectories, a subset of which _need_ to be terminated at a certain point to avoid getting too close to a singularity, then just setting an array to track the time doesn't avoid the singularity. In turn, this slows down the whole batch drastically (especially with adaptive time stepping), and/or causes all sorts of numerical errors. 

I'm not sure if supporting `terminate!` for ensemble array methods would be particularly efficient, since tracking the time is equivalent to the user truncating the solutions once they have them -- it wouldn't actually stop a given trajectory from continuing...
